### DISCUSSION TITLE
How does custom route work with auto api

### DISCUSSION THREAD

**Discussion by rakeshkgsg** [abpframework/abp#10286] [Q&A] [answered]
Am curious to know how auto api controllers work with custom routes especially in blazor web assembly.

The documentation at https://docs.abp.io/en/abp/latest/API/Auto-API-Controllers states that we can create a custom route as below :

```
Configure<AbpAspNetCoreMvcOptions>(options =>
{
    options.ConventionalControllers
        .Create(typeof(BookStoreApplicationModule).Assembly, opts =>
            {
                opts.RootPath = "myroute";
            });
});
```

So now the application services defined in the application module are automatically converted to api with /api/myroute. This is awesome and on the client we implement the dynamic client proxies as stated here https://docs.abp.io/en/abp/latest/API/Dynamic-CSharp-API-Clients

```
context.Services.AddHttpClientProxies(
            typeof(BookStoreApplicationContractsModule).Assembly
        );
```
My question is how does the client proxies know the custom routes of the auto apis. Or is there some additional configuration that needs to be done. Would love to understand the mechanism of how this works.

**Comment by realLiangshiwei** [comment, marked as answer]
Hi,

We have a get API definition endpoint: https://github.com/abpframework/abp/blob/dev/framework/src/Volo.Abp.AspNetCore.Mvc/Volo/Abp/AspNetCore/Mvc/ApiExploring/AbpApiDefinitionController.cs
